### PR TITLE
prune: Show which rule was applied to keep archive

### PR DESCRIPTION
**Pull request by NiklasMM**
Prune now shows for each kept archive:
 * Which rule is responsible for keeping this archive
 * How many archives have been kept by this rule so far

Ref #2886

Here's an example output:

```
Sun Oct 29 16:14:18 CET 2017 Pruning repository

Keeping archive (rule: secondly #1):     somehost-2017-10-29T16:14:17            Sun, 2017-10-29 16:14:17 [585c3f70354e0af201c7eb921a7bcf255178641712ca6f6a1ca1bfe3cf5668e8]
Pruning archive (1/1):                   somehost-2017-10-29T14:57:33            Sun, 2017-10-29 14:57:33 [9e9d5dd39948802af0f15c4ef19671f0ba90fad4bca134f62d9c5933acfde070]
Keeping archive (rule: daily #1):        somehost-2017-10-28T17:21:03            Sat, 2017-10-28 17:21:03 [b38fb5535b29e587b9332b8804bd9adc7a6a914b0910da4b8d392b8feb4305da]

```

### CODE REVIEW

File: src/borg/archiver.py [hunk 1]
```diff
@@ -1326,45 +1326,48 @@ def do_prune(self, args, repository, manifest, key):
         # that is newer than a successfully completed backup - and killing the successful backup.
         archives = [arch for arch in archives_checkpoints if arch not in checkpoints]
         keep = []
+        # collect the rule responsivle for the keeping of each archive in this dict
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: typo

File: src/borg/archiver.py [hunk 2]
```diff
@@ -1326,45 +1326,48 @@ def do_prune(self, args, repository, manifest, key):
...
+        # keys are archive ids, values are a tuple
+        #   (<rulename>, <how many archived kept by this rule so far >)
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: typo

**Comment by NiklasMM** [comment]
Fixed the typos :+1:

**Comment by ThomasWaldmann** [comment]
https://travis-ci.org/borgbackup/borg/jobs/294591619#L816

File: src/borg/testsuite/helpers.py [hunk 3]
```diff
@@ -333,40 +332,55 @@ def test(self):
...
+        ("daily", 3, (13, 8, 7)),
+        ("hourly", 3, (13, 10, 8)),
+        ("minutely", 3, (13, 10, 9)),
+        ("secondly", 4, (13, 12, 11, 10)),
+        ("daily", 0, [])
+    ]
+)
+def test_prune_split(rule, num_to_keep, expected_ids):
+    def subset(lst, ids):
+        return {i for i in lst if i.id in ids}
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/helpers.py]: guess there should be an empty line below this for pep8.

File: src/borg/testsuite/helpers.py [hunk 4]
```diff
@@ -333,40 +332,55 @@ def test(self):
...
+        MockArchive(datetime(2017, 10, 1, 10, 0, 1, tzinfo=timezone.utc), 11),
+        MockArchive(datetime(2017, 10, 1, 10, 0, 3, tzinfo=timezone.utc), 12),
+        MockArchive(datetime(2017, 10, 1, 10, 0, 5, tzinfo=timezone.utc), 13),
+    ]
+    kept_because = {}
+    keep = prune_split(archives, rule, num_to_keep, kept_because)
+
+    assert set(keep) == subset(archives, expected_ids)
+    for item in keep:
+        assert kept_because[item.id][0] == rule
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/helpers.py]: thanks for rewriting this!

File: src/borg/archiver.py [hunk 5]
```diff
@@ -1326,45 +1326,48 @@ def do_prune(self, args, repository, manifest, key):
...
-                        list_logger.info('Keeping archive: %s' % format_archive(archive))
+                    if is_checkpoint(archive.name):
+                        log_message = 'Keeping checkpoint archive:'
+                    else:
+                        log_message = 'Keeping archive (rule: {rule} #{num}):'.format(
+                            rule=kept_because[archive.id][0], num=kept_because[archive.id][1]
+                        )
+                if args.output_list:
+                    list_logger.info("{message} {archive}".format(
+                        message=log_message.ljust(40), archive=format_archive(archive)
```
[the remark this answers is not in the document]
Comment (NiklasMM) [reply src/borg/archiver.py]: yes, it does. Didn't know that one!

**Comment by NiklasMM** [comment]
Thanks for the review @ThomasWaldmann , I have made the requested adjustments, squashed everything together and rebased.

**Pull request merged**